### PR TITLE
Revert "Revert "[Data] Reduce internal Ray Data stack trace output by default""

### LINKED ISSUE #40802: [data] Capture Data-internal exceptions
### Description

Currently Data throws all exceptions to the user, both from user-defined code and Data-internal code. Errors are not differentiated, which makes it difficult to understand the error output.

Ideally, we should capture exceptions from user-defined code, throw these as a separate type of error, and match the stacktrace to the original user code. Also, we should improve the error message for Data-internal errors to direct the user to file a GitHub issue.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by scottjlee**
Reverts ray-project/ray#43638, adding changes from original PR https://github.com/ray-project/ray/pull/43251 with fix for breaking osx / windows tests.

Closes https://github.com/ray-project/ray/issues/40802

Postmerge test run: https://buildkite.com/ray-project/postmerge/builds/3306

### CODE REVIEW

**Comment by scottjlee** [comment]
The fix / new changes are in https://github.com/ray-project/ray/pull/43735/commits/03e063520809350635a6094f2d9ed4ac467121b1

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -1025,6 +1025,10 @@ cdef c_bool determine_if_retryable(
     # Python API should have converted the list of exceptions to a tuple.
     assert isinstance(exception_allowlist, tuple)
 
+    # For exceptions raised when running UDFs in Ray Data, we need to unwrap the special
+    # exception type thrown by Ray Data in order to get the underlying exception.
+    if isinstance(e, ray.data.exceptions.UserCodeException):
```
Comment (raulchen) [inline python/ray/_raylet.pyx:1030]:
```suggestion
    if isinstance(e, ray.exceptions.UserCodeException):
```

File: python/ray/exceptions.py [hunk 2]
```diff
@@ -350,6 +350,15 @@ def __str__(self):
         return error_msg
 
 
+@DeveloperAPI
+class UserCodeException(RayError):
+    """Indicates that an exception occurred while executing user code.
+    Used to catch user code exceptions which arise from executing
+    Ray Data code."""
```
Comment (raulchen) [inline python/ray/exceptions.py:357]: As it's a ray core level thing, better to say something more general here. for example, allow library developers to wrap user code exceptions from a remote task or actor. The `retry_exceptions` flag will still respect the cause of this exception.

**Review by raulchen — APPROVED** [review]

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (scottjlee) [reply python/ray/_raylet.pyx:1030]: thanks, forgot the most important change 😅 also kicking off postmerge tests to ensure there is no issue.

**Review by raulchen — APPROVED** [review]

**Pull request merged**